records/2023-03-28.secret-management.md

```diff
@@ -0,0 +1,189 @@
+
+There are multiple types of Secret Engines. These types typically dictate where the secrets are stored. The default secret engine is a V2 KeyVault which hosts the data within the Vault instance. This is usually sufficient, however it is possible to leverage other Secret Engines, and Vault has Secret Engines that leverage AWS Secrets Manager, and Azure Key Vault as the back end storage repositoroes. Further investigation could be done to see if integrating Vault with AWS Secrets Manager provides the best of both worlds for Notify. 
+
+Secrets and secret key value pairs can also be configured to perform Just In Time (JIT) authentication. In this scenario, when an application needs access to a specific resource, it will make a call to vault and request a JIT token which will have a very short TTL, creating what is essentially a one time use key. This is highly beneficial for security but requires considerable set up, and may have performance impacts. That being said, it is something worth investigating in the future. 
+
+#### IAM 
+
+Vault IAM is similar to AWS IAM, which is to say (unfortunately) that it can be rather complex. Vault identifies user and system accounts as "Entities". These entities are identified as a GUID, making it difficult at a glance to see exactly what an entity is. For this reason, Vault ties a secondary identification method to these entities called "Aliases". Typically the alias of an entity is the display name of a user or system account. Unfortunately, in most IAM scenarios it is not possible to administer an entity based on its alias. Instead once the entity is found by the alias, it is required to make note of the entity guid to manage it.
+
+A positive of vault is that it is very flexible in the authentication methods that can be used. The default vault root administrator will use temporary token authentication, but this is not the recommended approach for day to day use. Instead, authentication can be done via standard username/password, OIDC login to an external IDP (GitHub etc), AWS account delegation, Kubernetes, and more. This makes it very easy to integrate virtually any system within vault. Vault uses the concept of roles that are assigned to an authentication method. The role is a group of policies and permissions that can be assigned to a user who authenticates with this auth method. There will be a default role that will be assigned to the auth method, but additional roles can be specified and requested during login.
+
+Vault uses policies for permission management. A policy is a JSON structured document that defines the permissions to the individual secret engines. Policies use keywords such as "Read","Write" and "List" to define how an entity is permitted to interact with a specific secret engine. These policies can be signed directly to an entity, or assigned to a group (which is a group of entities), or assigned to a role (which is a collection of policies and groups). 
```

Comment:
That kinda answer my previous question about scope of permissions if I get this right. So yes, we can isolate permissions to engine, but it might not for finer grained permissions with that engine. Correct?

Reply:
djasfYou can lock it down to a single secret, but keepinfg in mind that a secret has multiple key/value pairs, I don't believe you can lock it down to a single kv pair within that secret.

Secret: /notifysecrets/production/api
kv pairs:
apikey=1231241
databasepassword=password

For example:
Locking down to: /notifysecrets/production/api is possible, but the policy will have access to all of the key value pairs in there.